File: xis-javascript/src/main/js/classes/connect/ServerResponse.js
/**
 * @property {Number} status
 * @property {any} data
 * @property {string} nextURL
 * @property {string} nextFrontletId
 * @property {string} nextModalId
 * @property {boolean} closeModal
 * @property {boolean} reloadModalParent
 * @property {any} formData
 * @property {Array<string>} returnedFormDataKeys
 * @property {any} localStorageData
 * @property {any} localDatabaseData
 * @property {any} globalVariableData
 * @property {Object} frontletParameters
 * @property {Object} modalParameters
 * @property {string} frontletContainerId
 * @property {Array} reloadFrontlets
 * @property {Array} updateEventKeys
 * @property {any} sessionStorageData
 * @property {any} clientStateData
 * @property {any} validatorMessages
 * @property {boolean} authenticated
 * @property {Array<string>} userRoles
 * @property {Object} idVariables
 * @property {boolean} reloadPage
 * @property {string} redirectUrl
 * @property {string} actionProcessing
 * @property {string} annotatedTitle
 * @property {string} annotatedAddress
 * @property {Array} defaultFrontlets
 */

class ServerResponse {

    constructor() {
        this.status = -1;
        this.data = {};
        this.nextURL = '';
        this.nextFrontletId = '';
        this.nextModalId = '';
        this.closeModal = false;
        this.reloadModalParent = false;
        this.formData = {};
        this.returnedFormDataKeys = [];
        this.localStorageData = {};
        this.globalVariableData = {};
        this.localDatabaseData = {};
        this.frontletParameters = {};
        this.modalParameters = {};
        this.frontletContainerId = '';
        this.reloadFrontlets = [];
        this.sessionStorageData = {};
        this.clientStateData = {};
        this.validatorMessages = { };
        this.authenticated = false;
        this.userRoles = [];
        this.idVariables = {};
        this.reloadPage = false;
        this.redirectUrl = '';
        this.actionProcessing = 'NONE';
        this.updateEventKeys = [];
        this.annotatedTitle = undefined;
        this.annotatedAddress = undefined;
        this.defaultFrontlets = [];
    }


    containsGlobals() {
       if (this.globalVariableData && Object.keys(this.globalVariableData).length > 0) {
           return true;
       }
       if (this.localStorageData && Object.keys(this.localStorageData).length > 0) {
           return true;
       }
       if (this.localDatabaseData && Object.keys(this.localDatabaseData).length > 0) {
           return true;
       }
       if (this.sessionStorageData && Object.keys(this.sessionStorageData).length > 0) {
           return true;
       }
       if (this.clientStateData && Object.keys(this.clientStateData).length > 0) {
           return true;
       }
       return false;
    }
}

File: xis-adapter-spi/src/main/java/one/xis/server/ServerResponse.java
package one.xis.server;


import com.fasterxml.jackson.annotation.JsonInclude;
import com.google.gson.annotations.JsonAdapter;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import one.xis.validation.ValidatorMessages;

import java.util.*;


@Data
@AllArgsConstructor
@NoArgsConstructor
@JsonInclude(JsonInclude.Include.NON_DEFAULT)
@JsonAdapter(ServerResponseTypeAdapter.class)

public class ServerResponse {

    private transient int status;
    private Map<String, Object> data = new HashMap<>();
    private Map<String, String> idVariables = new HashMap<>();
    private Map<String, Object> formData = new HashMap<>();
    private Collection<String> returnedFormDataKeys = new HashSet<>();
    private Map<String, Object> localStorageData = new HashMap<>();
    private Map<String, Object> localDatabaseData = new HashMap<>();
    private Map<String, Object> sessionStorageData = new HashMap<>();
    private Map<String, Object> clientStorageData = new HashMap<>();
    private Map<String, Object> frontletParameters = new HashMap<>();
    private Map<String, Object> modalParameters = new HashMap<>();
    private boolean authenticated;
    private Collection<String> userRoles = new HashSet<>();
    private Collection<String> updateEventKeys = new ArrayList<>();
    private Collection<String> reloadFrontlets = new ArrayList<>(); // TODO remove?
    private String nextURL;
    private String nextFrontletId;
    private String nextModalId;
    private boolean closeModal;
    private boolean reloadModalParent;
    private ValidatorMessages validatorMessages = new ValidatorMessages();
    private boolean reloadPage; // TODO do we need this?
    private String frontletContainerId;
    private String redirectUrl;
    private ActionProcessing actionProcessing = ActionProcessing.NONE;
    private String annotatedTitle;
    private String annotatedAddress;
    private Collection<DefaultFrontlet> defaultFrontlets = new HashSet<>();

    void clear() {
        // do not clear store data
        status = 0;
        data.clear();
        formData.clear();
        returnedFormDataKeys.clear();
        reloadFrontlets.clear();
        frontletParameters.clear();
        modalParameters.clear();
        nextURL = null;
        nextFrontletId = null;
        nextModalId = null;
        closeModal = false;
        reloadModalParent = false;
        validatorMessages = new ValidatorMessages();
        reloadPage = false;
        frontletContainerId = null;
        defaultFrontlets.clear();
    }
}
